File: spikes/wasm-mat4/mat4.mjs
// Single file loader for the mat4 SIMD kernels.
// The module is inlined as base64, so there is no fetch, no bundler plugin and
// no async anywhere in the API. Works the same in node, a bundler and from a CDN.

const WASM_BASE64 =
    'AGFzbQEAAAABFgNgBH9/f38AYAN/f38AYAV/f39/fwADBgUAAQAAAgUEAQCAAgYIAX8BQYCIBAsHQgUGbWVtb3J5AgAJbXVs' +
    'X2JhdGNoAAANbXVsX2Jyb2FkY2FzdAACCWhpZXJhcmNoeQADDWNvbXBvc2VfYmF0Y2gABArLCAVGACADQQAgA0EAShshAwNA' +
    'AkAgAw0ADwsgACABIAIQgYCAgAAgA0F/aiEDIABBwABqIQAgAUHAAGohASACQcAAaiECDAALC8cBAQV7IAH9AAAwIQMgAf0A' +
    'ACAhBCAB/QAAECEFIAH9AAAAIQZBACEBA0ACQCABQcAARw0ADwsgACABaiAGIAIgAWr9AAAAIgcgB/0NAAECAwABAgMAAQID' +
    'AAECA/3mASAFIAcgB/0NBAUGBwQFBgcEBQYHBAUGB/3mAf3kASAEIAcgB/0NCAkKCwgJCgsICQoLCAkKC/3mASADIAcgB/0N' +
    'DA0ODwwNDg8MDQ4PDA0OD/3mAf3kAf3kAf0LAAAgAUEQaiEBDAALC/4BAgF/BXtBACEEIANBACADQQBKGyEDIAH9AAAwIQUg' +
    'Af0AACAhBiAB/QAAECEHIAH9AAAAIQgCQANAIAQgA0YNAUEAIQEDQAJAIAFBwABHDQAgAkHAAGohAiAAQcAAaiEAIARBAWoh' +
    'BAwCCyAAIAFqIAggAiABav0AAAAiCSAJ/Q0AAQIDAAECAwABAgMAAQID/eYBIAcgCSAJ/Q0EBQYHBAUGBwQFBgcEBQYH/eYB' +
    '/eQBIAYgCSAJ/Q0ICQoLCAkKCwgJCgsICQoL/eYBIAUgCSAJ/Q0MDQ4PDA0ODwwNDg8MDQ4P/eYB/eQB/eQB/QsAACABQRBq' +
    'IQEMAAsLCwulAQEEf0EAIQQgA0EAIANBAEobIQUgACEGIAEhBwJAA0AgBCAFRg0BAkACQCACIARBAnRqKAIAIgNBf0oNAEEA' +
    'IQMDQCADQcAARg0CIAYgA2ogByADav0AAAD9CwAAIANBEGohAwwACwsgACAEQQZ0aiAAIANBBnRqIAEgBEEEdEECdGoQgYCA' +
    'gAALIAZBwABqIQYgB0HAAGohByAEQQFqIQQMAAsLC5EDAwJ/A3sLfUEAIQUgBEEAIARBAEobIQQDQAJAIAQNAA8LIAMgBWoi' +
    'Bv0JAgAhByAGQQRq/QkCACEIIABBIGr9DAAAAAAAAAAAAAAAAAAAAAAiCSACKgIAIgogAkEIaioCACILIAuSIgyUIg0gAkEE' +
    'aioCACIOIA6SIg8gAkEMaioCACIQlCIRkv0gACAOIAyUIhIgCiAKkiITIBCUIhST/SABQwAAgD8gCiATlCITIA4gD5QiDpKT' +
    '/SACIAZBCGr9CQIA/eYB/QsAACAAQRBqIAggCSAKIA+UIgogECAMlCIPk/0gAEMAAIA/IBMgCyAMlCIMkpP9IAEgEiAUkv0g' +
    'Av3mAf0LAAAgACAHIAlDAACAPyAOIAySk/0gACAKIA+S/SABIA0gEZP9IAL95gH9CwAAIABBMGr9DAAAAAAAAAAAAAAAAAAA' +
    'gD8gASAFaiIGKgIA/SAAIAZBBGoqAgD9IAEgBkEIaioCAP0gAv0LAAAgBEF/aiEEIABBwABqIQAgBUEMaiEFIAJBEGohAgwA' +
    'Cws=';

function decode(base64) {
    if (typeof Buffer !== 'undefined') return Buffer.from(base64, 'base64');
    const bin = atob(base64);
    const bytes = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; i++) bytes[i] = bin.charCodeAt(i);
    return bytes;
}

/**
 * True when this engine can run this module.
 *
 * Validates the real module rather than a stand in probe, so the check tests
 * exactly the features the kernels use and cannot drift away from them.
 */
export function supported() {
    return WebAssembly.validate(decode(WASM_BASE64));
}

/**
 * Creates a kernel instance owning `capacity * 16` floats of scratch per buffer.
 *
 * Memory is sized once and never grows, because `memory.grow` detaches every
 * view already handed out. Allocate to capacity up front.
 */
export function createMat4Wasm(capacity) {
    const bytes = decode(WASM_BASE64);
    // sync compile, legal on the main thread only while the module stays under 4KB
    const instance = new WebAssembly.Instance(new WebAssembly.Module(bytes), {});
    const memory = instance.exports.memory;

    const pages = Math.ceil((capacity * 64 * 4 + 65536) / 65536);
    if (pages > memory.buffer.byteLength / 65536) memory.grow(pages - memory.buffer.byteLength / 65536);

    let offset = 0;
    const own = (view) => {
        if (view.buffer !== memory.buffer)
            throw new Error('buffer was not allocated by this module, so it is not resident');
        return view.byteOffset;
    };
    const bump = (bytesNeeded) => {
        const at = offset;
        offset += (bytesNeeded + 15) & ~15; // keep every allocation 16 byte aligned
        if (offset > memory.buffer.byteLength) throw new Error('out of wasm memory, raise capacity');
        return at;
    };

    return {
        memory,
        supported: supported(),

        /** A flat array of `count` matrices, 16 floats each, resident in wasm memory. */
        allocMat4: (count) => new Float32Array(memory.buffer, bump(count * 64), count * 16),
        /** A flat array of `count` vec3 or quat, for the compose inputs. */
        allocFloat: (count, stride) => new Float32Array(memory.buffer, bump(count * stride * 4), count * stride),
        /** Parent indices for `hierarchy`, negative meaning root. */
        allocIndex: (count) => new Int32Array(memory.buffer, bump(count * 4), count),

        /** out[i] = a[i] * b[i] */
        multiply(out, a, b, count = out.length / 16) {
            instance.exports.mul_batch(own(out), own(a), own(b), count);
            return out;
        },
        /** out[i] = m * a[i], for viewProjection times every world matrix */
        multiplyBroadcast(out, m, a, count = out.length / 16) {
            instance.exports.mul_broadcast(own(out), own(m), own(a), count);
            return out;
        },
        /** world[i] = world[parent[i]] * local[i]. Parents must precede children. */
        hierarchy(world, local, parent, count = world.length / 16) {
            instance.exports.hierarchy(own(world), own(local), own(parent), count);
            return world;
        },
        /** out[i] = translate(pos[i]) * rotate(quat[i]) * scale(scale[i]) */
        compose(out, pos, quat, scale, count = out.length / 16) {
            instance.exports.compose_batch(own(out), own(pos), own(quat), own(scale), count);
            return out;
        },
    };
}

File: spikes/wasm-mat4/mat4.c
#include <wasm_simd128.h>

#define EXPORT __attribute__((visibility("default")))

// broadcast lane l of v across all four lanes
#define BC(v, l) wasm_i32x4_shuffle(v, v, l, l, l, l)

// column major 4x4, out_col_j = sum_k a_col_k * b[j*4+k]
static inline void mul4(float *out, const float *a, const float *b) {
    v128_t a0 = wasm_v128_load(a);
    v128_t a1 = wasm_v128_load(a + 4);
    v128_t a2 = wasm_v128_load(a + 8);
    v128_t a3 = wasm_v128_load(a + 12);

    // tree reduction, two dependent adds rather than three
    for (int j = 0; j < 4; j++) {
        v128_t bj = wasm_v128_load(b + j * 4);
        v128_t lo = wasm_f32x4_add(wasm_f32x4_mul(a0, BC(bj, 0)), wasm_f32x4_mul(a1, BC(bj, 1)));
        v128_t hi = wasm_f32x4_add(wasm_f32x4_mul(a2, BC(bj, 2)), wasm_f32x4_mul(a3, BC(bj, 3)));
        wasm_v128_store(out + j * 4, wasm_f32x4_add(lo, hi));
    }
}

// n pairwise multiplies
EXPORT void mul_batch(float *out, const float *a, const float *b, int n) {
    for (int i = 0; i < n; i++) mul4(out + i * 16, a + i * 16, b + i * 16);
}

// one matrix times n matrices, the viewProjection * world workload
EXPORT void mul_broadcast(float *out, const float *m, const float *a, int n) {
    v128_t m0 = wasm_v128_load(m);
    v128_t m1 = wasm_v128_load(m + 4);
    v128_t m2 = wasm_v128_load(m + 8);
    v128_t m3 = wasm_v128_load(m + 12);

    for (int i = 0; i < n; i++) {
        const float *a_i = a + i * 16;
        float *o = out + i * 16;
        for (int j = 0; j < 4; j++) {
            v128_t aj = wasm_v128_load(a_i + j * 4);
            v128_t lo = wasm_f32x4_add(wasm_f32x4_mul(m0, BC(aj, 0)), wasm_f32x4_mul(m1, BC(aj, 1)));
            v128_t hi = wasm_f32x4_add(wasm_f32x4_mul(m2, BC(aj, 2)), wasm_f32x4_mul(m3, BC(aj, 3)));
            wasm_v128_store(o + j * 4, wasm_f32x4_add(lo, hi));
        }
    }
}

// scene graph propagation, parents must precede children
EXPORT void hierarchy(float *world, const float *local, const int *parent, int n) {
    for (int i = 0; i < n; i++) {
        int p = parent[i];
        if (p < 0) {
            for (int j = 0; j < 4; j++)
                wasm_v128_store(world + i * 16 + j * 4, wasm_v128_load(local + i * 16 + j * 4));
        } else {
            mul4(world + i * 16, world + p * 16, local + i * 16);
        }
    }
}

// TRS compose from position, quaternion, scale, the instancing workload
EXPORT void compose_batch(float *out, const float *pos, const float *quat, const float *scale, int n) {
    for (int i = 0; i < n; i++) {
        float x = quat[i * 4], y = quat[i * 4 + 1], z = quat[i * 4 + 2], w = quat[i * 4 + 3];
        float x2 = x + x, y2 = y + y, z2 = z + z;
        float xx = x * x2, xy = x * y2, xz = x * z2;
        float yy = y * y2, yz = y * z2, zz = z * z2;
        float wx = w * x2, wy = w * y2, wz = w * z2;
        float sx = scale[i * 3], sy = scale[i * 3 + 1], sz = scale[i * 3 + 2];

        float *o = out + i * 16;
        wasm_v128_store(o, wasm_f32x4_mul(wasm_f32x4_make(1 - (yy + zz), xy + wz, xz - wy, 0), wasm_f32x4_splat(sx)));
        wasm_v128_store(o + 4, wasm_f32x4_mul(wasm_f32x4_make(xy - wz, 1 - (xx + zz), yz + wx, 0), wasm_f32x4_splat(sy)));
        wasm_v128_store(o + 8, wasm_f32x4_mul(wasm_f32x4_make(xz + wy, yz - wx, 1 - (xx + yy), 0), wasm_f32x4_splat(sz)));
        wasm_v128_store(o + 12, wasm_f32x4_make(pos[i * 3], pos[i * 3 + 1], pos[i * 3 + 2], 1));
    }
}
